weaviate_txtai/ann/weaviate.py

```diff
@@ -0,0 +1,156 @@
+                return True
+
+        return False
+
+    def _create_schema(self):
+        schema = self.weaviate_config.get("schema", DEFAULT_SCHEMA)
+
+        if not self._is_valid_schema(schema):
+            raise weaviate.exceptions.SchemaValidationException(
+                f"Class {schema['class']} must have a property named 'docid' of type 'int'"
+            )
+
+        if self.client.schema.contains(schema):
+            if not self.overwrite_index:
+                raise weaviate.exceptions.ObjectAlreadyExistsException(
+                    f"Index {schema['class']} already exists"
+                )
+            else:
+                self.client.schema.delete_class(schema["class"])
+
+        self.client.schema.create_class(schema)
+        self.index_name = schema["class"]
+
+    def index(self, embeddings):
+        self.append(embeddings)
+
+    def append(self, embeddings):
+
+        with self.client.batch as batch:
+            for embedding in embeddings:
+                random_identifier = uuid.uuid4()
+                object_uuid = generate_uuid5(random_identifier)
+                batch.add_data_object(
+                    data_object={
+                        "docid": self.config["offset"],
+                    },
+                    class_name=self.index_name,
+                    vector=embedding,
+                    uuid=object_uuid,
+                )
+
+                self.config["offset"] += 1
+
+    def _get_uuid_from_docid(self, docid):
+
+        results = (
+            self.client.query.get(self.index_name)
+            .with_additional("id")
+            .with_where(
+                {
+                    "path": ["docid"],
+                    "operator": "Equal",
+                    "valueInt": docid,
+                }
+            )
+            .do()
+        )
+
+        return results["data"]["Get"][self.index_name][0]["_additional"]["id"]
+
+    def delete(self, ids):
+
+        for id in ids:
+            # TODO: rewrite when weaviate supports IN operator
+            #       See: https://github.com/weaviate/weaviate/issues/2387
+            uuid = self._get_uuid_from_docid(id)
+            self.client.data_object.delete(uuid, class_name=self.index_name)
+
+    def search(self, queries, limit):
+
+        nearVector = {"vector": queries[0]}
+
+        # use distance to score similarity
+        # TODO: discuss if this is the best way to do it
+        #       in txtai, higher similarity score means more similar
+        #       but in weaviate, lower distance means more similar
+        results = (
+            self.client.query.get(self.index_name, properties=["docid"])
+            .with_additional("distance")
+            .with_near_vector(nearVector)
+            .with_limit(limit)
+            .do()
+        )
+
+        results = results["data"]["Get"][self.index_name]
+
+        return [
+            [(result["docid"], result["_additional"]["distance"]) for result in results]
```

Comment:
certainty is [only available if the distance metric is cosine](https://weaviate.io/developers/weaviate/api/graphql/additional-properties#certainty-only-for-cosine-distance).

what if the user wants to try another distance metrice?

Reply:
@davidmezzetti weaviate has a few [distance metrics](https://weaviate.io/developers/weaviate/configuration/distances) and none of them bounded between 0 and 1. 

What is the default similarity score used in txtai?

Reply:
txtai uses cosine similarity on normalized vectors. It filters out negative scores to bound from 0-1 (like what you ran into in your testing).

Reply:
@davidmezzetti thanks for clarifying. Since txtai always uses cosine similarity, I will add a check to the weaviate schema to make sure it the distance metric is cosine.